cpp/src/community/leiden_impl.cuh

```diff
@@ -42,17 +45,508 @@ void check_clustering(graph_view_t<vertex_t, edge_t, false, multi_gpu> const& gr
+                                dst_louvain_assignment_cache,
+                                up_down);
+
+    // Clear buffer and contract the graph
+
+    cluster_keys.resize(0, handle.get_stream());
+    cluster_weights.resize(0, handle.get_stream());
+    vertex_weights.resize(0, handle.get_stream());
+    louvain_assignment_for_vertices.resize(0, handle.get_stream());
+    cluster_keys.shrink_to_fit(handle.get_stream());
+    cluster_weights.shrink_to_fit(handle.get_stream());
+    vertex_weights.shrink_to_fit(handle.get_stream());
+    louvain_assignment_for_vertices.shrink_to_fit(handle.get_stream());
+    src_vertex_weights_cache.clear(handle);
+    src_louvain_assignment_cache.clear(handle);
+    dst_louvain_assignment_cache.clear(handle);
+
+#ifdef TIMING
+    detail::timer_start<graph_view_t::is_multi_gpu>(handle, hr_timer, "contract graph");
+#endif
+    // Create aggregate graph based on refined (leiden) partition
+
+    std::optional<rmm::device_uvector<vertex_t>> cluster_assignment{std::nullopt};
+
+    // prev_nr_of_vertices = current_graph_view.number_of_vertices();
+    std::tie(current_graph, coarsen_graph_edge_property, cluster_assignment) =
+      cugraph::detail::graph_contraction(
+        handle,
+        current_graph_view,
+        edge_weight_view,
+        raft::device_span<vertex_t>{refined_leiden_partition.begin(),
+                                    refined_leiden_partition.size()});
+    current_graph_view = current_graph.view();
+
+    edge_weight_view = std::make_optional<edge_property_view_t<edge_t, weight_t const*>>(
+      (*coarsen_graph_edge_property).view());
+
+    // After call to relabel, cluster_assignment contains louvain partition of the aggregated graph
+    relabel<vertex_t, multi_gpu>(
+      handle,
+      std::make_tuple(static_cast<vertex_t const*>(leiden_to_louvain_map.first.begin()),
+                      static_cast<vertex_t const*>(leiden_to_louvain_map.second.begin())),
+      leiden_to_louvain_map.first.size(),
+      (*cluster_assignment).data(),
+      (*cluster_assignment).size(),
+      false);
+
+    louvain_of_refined_partition.resize(current_graph_view.local_vertex_partition_range_size(),
+                                        handle.get_stream());
+
+    raft::copy(louvain_of_refined_partition.begin(),
+               (*cluster_assignment).begin(),
+               (*cluster_assignment).size(),
+               handle.get_stream());
+
+    if (cluster_assignment) {
+      (*cluster_assignment).resize(0, handle.get_stream());
+      (*cluster_assignment).shrink_to_fit(handle.get_stream());
+    }
+
+#ifdef TIMING
+    detail::timer_stop<graph_view_t::is_multi_gpu>(handle, hr_timer);
+#endif
+  }
+
+#ifdef TIMING
+  detail::timer_display<graph_view_t::is_multi_gpu>(handle, hr_timer, std::cout);
+#endif
+
+  return std::make_pair(std::move(dendrogram), best_modularity);
 }
 
 // FIXME: Can we have a common flatten_dendrogram to be used by both
```

Comment:
Sure.  You could move this function to `detail/common_methods.{app,cuh,sg,mg}` and have both Louvain and Leiden implementations call it.  This can be deferred to later if we're running out of time.